### PR TITLE
HHH-11409 Registered collection binding

### PR DESCRIPTION
**Pull request by coladict**
Bind registered collection types using their type handler.

Also I accidentally found a bug in org.hibernate.query.internal.AbstractProducedQuery.setParameter(int position, Object value) near line 495. Should be `else if`, instead of just `if`.

### CODE REVIEW

**Comment by vladmihalcea** [comment]
Thanks for your Pull Request. Can you provide a test case to replicate the issue?

**Comment by coladict** [comment]
Yes, https://github.com/mopano/hibernate-json-type/tree/binding-test

Hibernate build passes all the old tests when running `./gradlew clean build publishToMavenLocal` and the contributor build's `com.mopano.hibernate.test.JsonContributionTest.testParameter` passes the binding of JsonArray types. All three implementations of JSR-353 I could find in Maven Central extend `java.util.AbstractList` for that type, which caused it to be unable to bind as a parameter.

**Comment by vladmihalcea** [comment]
I added a new commit trying to replicate your issue, but I couldn't. Check out [my commit](https://github.com/hibernate/hibernate-orm/commit/d9e5edc9c0268c25d780c78d97a7c918b1b6a9ae), and adjust it so that you can replicate the issue. If you can;t replicate it, I'll close the Jira issue.

**Comment by coladict** [comment]
I added a unit test to the main project PR. It also serves as a test for the `org.hibernate.cfg.AvailableSettings.CLASSLOADERS` setting and for `TypeContributor`s in general. Everything passes with `clean build publishToMavenLocal`, thought I had to use a native query, because the parser thinks the parameter handler is org.hibernate.type.StringType. Apparently it checks the database type, rather than the field class. Apparently that's a bug that's been there and no one has run into and reported before. Though I don't think it's worth fixing before 6.0, because it will probably require a major overhaul of some things.

**Comment by vladmihalcea** [comment]
Did you sign the [CLA](https://cla.jboss.org/)? I need you to sign it to integrate your Pull Request.

**Comment by coladict** [comment]
[Yes](http://i.imgur.com/dOQ7Tu1.png).

File: hibernate-core/src/test/java/org/hibernate/jpa/test/query/QueryTest.java [hunk 1]
```diff
@@ -1451,4 +1483,41 @@ public void testGetSingleResultWithNoResultException() {
...
+
+			em.getTransaction().begin();
+			LongList ll = new LongList();
+			ll.add( 5L );
+			ll.add( 11L );
+			ll.add( 6123L );
+			ll.add( -61235L );
+			ll.add( 24L );
+			Query tq = em.createNativeQuery( "SELECT * FROM Item WHERE descr = ?", Item.class );
+			tq.setParameter(1, ll);
```
Comment (vladmihalcea) [inline hibernate-core/src/test/java/org/hibernate/jpa/test/query/QueryTest.java]: This is wrong. The Java type is `String` so the parameter should be set as:

    .setParameter(1, "[5,11,6123,-61235,24]")

in which case it works just fine even after I revert your changes to `AbstractProducedQuery`.

Your test should use the `LongList` as an entity attribute for the `TypeContributor` to have a purpose in this test, right?

**Comment by vladmihalcea** [comment]
There's an issue in your test. Check out my comment.

**Comment by coladict** [comment]
> This is wrong. The Java type is String so the parameter should be set as...

In my remark about `org.hibernate.type.StringType` and type resolution through query parsing I forgot that the entity I'm using isn't a custom one that contains a `LongList` element. I took that shortcut as to not make a new entity type. I could take the time and move it to a separate test with it's own entity type. Make it a more complete TypeContributer test, and not merely a test for the parameter feature.

**Comment by coladict** [comment]
The new `org.hibernate.test.type.ContributorTest` makes the goal more clear. Being able to bind any registered type to a query parameter, even if it's a collection.
Now both in native and JPQL queries.

It also serves as a test for the `hibernate.classLoaders` configuration property.

**Comment by vladmihalcea** [comment]
Refactor the tests and applied upstream. Thanks.

**Comment by coladict** [comment]
I'm kind of curious about this convention to apply it via a new commit with same author, but not committer. Why use it when it gives the false impression of a rejected solution in Jira?

**Comment by vladmihalcea** [comment]
It's not a new commit. It's what git does when you rebase/squash all cimmits and ammend my changes on top of your work. Also, in Git you also have[ author and committer](https://github.com/hibernate/hibernate-orm/commit/f8570017dfc22b33b494d885e1896a8c73d3489c) for this purpose:

> coladict committed with vladmihalcea on Jan 19

**Pull request closed without merging**